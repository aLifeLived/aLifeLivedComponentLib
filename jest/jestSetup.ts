/* eslint-disable no-undef */
jest.useFakeTimers({ legacyFakeTimers: true }); // Stops reference error from appearing
