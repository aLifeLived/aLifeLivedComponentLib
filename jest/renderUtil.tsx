import React from 'react';
import { ThemeProvider } from '../src/themes/ThemeProvider';
import { render as rtlRender } from '@testing-library/react-native';
import { library } from '@fortawesome/fontawesome-svg-core';
import { iconList } from '../src/themes/systemIcon/IconList';

library.add(iconList);

function render(ui: any, { ...renderOptions } = {}) {
  function Wrapper({ children }: { children: any }) {
    return <ThemeProvider>{children}</ThemeProvider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything for convenience
export * from '@testing-library/react-native';

// override render method
export { render };
