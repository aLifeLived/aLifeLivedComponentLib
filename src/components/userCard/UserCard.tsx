import React from 'react';
import { Avatar } from '../avatar/Avatar';
import { Box } from '../../themes/box/Box';
import { SystemButton } from '../../themes/systemButton/SystemButton';
import { Text } from '../../themes/text/Text';

type UserCardProps = {
  username: string;
  avatarUrl: string;
  onPress: () => void;
};

export const UserCard: React.FC<UserCardProps> = ({
  username,
  avatarUrl,
  onPress,
}) => {
  return (
    <SystemButton onPress={() => onPress()} width="100%">
      <Box flexDirection="row" alignItems="center">
        <Avatar
          iconHeight={50}
          iconWidth={50}
          source={{ uri: avatarUrl, priority: 'high' }}
          variant="roundAvatar"
        />
        <Box marginLeft="s" width="80%">
          <Text variant="body" numberOfLines={1}>
            {username}
          </Text>
        </Box>
      </Box>
    </SystemButton>
  );
};
