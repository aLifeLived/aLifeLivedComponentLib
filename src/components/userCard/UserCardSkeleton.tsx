import React from 'react';

// Components
import { Box } from '../../themes/box/Box';
import { SkeletonElement } from '../../themes/loadingSkeletons/SkeletonElement';
import { SystemAnimatableBox } from '../../themes/systemAnimatableBox/SystemAnimatableBox';

type UserCardSkeleton = {
  testID?: string;
};

export const UserCardSkeleton: React.FC<UserCardSkeleton> = ({ testID }) => {
  return (
    <SystemAnimatableBox
      flexDirection="row"
      animation="flash"
      iterationCount="infinite"
      alignItems="center"
      useNativeDriver={true}
    >
      <SkeletonElement variant="avatar" />
      <Box width="100%" marginLeft="s" testID={testID}>
        <SkeletonElement variant="title" />
      </Box>
    </SystemAnimatableBox>
  );
};
