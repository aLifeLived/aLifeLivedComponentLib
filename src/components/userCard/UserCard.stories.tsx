import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { action } from '@storybook/addon-actions';

// Component
import { UserCard } from './UserCard';
import { UserCardSkeleton } from './UserCardSkeleton';
import { Box } from '../../themes/box/Box';
import { SystemFlatList } from '../../themes/systemFlatList/SystemFlatList';

const data = [
  {
    id: '5f92cdce0cf217478ba93163',
    username: 'johnSmithWithAReallyReallyLongNameSoLOng',
    avatarImageMediaAsset: {
      url: 'https://picsum.photos/id/237/200/300',
    },
  },
  {
    id: '5f92cdce0cf217478ba93263',
    username: 'johnSmith',
    avatarImageMediaAsset: undefined,
  },
  {
    id: '5f92cdce0cf217478ba93663',
    username: 'johnSmith',
    avatarImageMediaAsset: {
      url: 'https://picsum.photos/id/237/200/300',
    },
  },
  {
    id: '5f92cdce0cf217478ba93561',
    username: 'johnSmith',
    avatarImageMediaAsset: {
      url: '',
    },
  },
];

storiesOf('User card', module)
  .addDecorator(Story => Story())
  .add('Default user card', () => (
    <Box
      backgroundColor="surface"
      flex={1}
      alignContent="center"
      alignItems="flex-start"
    >
      <SystemFlatList
        data={data}
        width="100%"
        keyExtractor={item => item.id}
        renderItem={({ item }) => (
          <Box marginTop="m" width="100%" paddingLeft="s">
            <UserCard
              onPress={action('On press fired')}
              username={item.username}
              avatarUrl={item.avatarImageMediaAsset?.url}
            />
          </Box>
        )}
      />
    </Box>
  ))
  .add('User card loading', () => (
    <Box margin="m">
      <UserCardSkeleton />
    </Box>
  ));
