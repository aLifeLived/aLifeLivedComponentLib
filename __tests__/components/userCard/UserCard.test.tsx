import React from 'react';
import { render, fireEvent } from './../../../jest/renderUtil';

// Component
import { UserCard } from '../../../src/components/userCard/UserCard';

const onPress = jest.fn();
const username = 'testUsername';
let avatarUrl = '';
const setup = (isUserSelected = false) =>
  render(
    <UserCard
      username={username}
      avatarUrl={avatarUrl}
      onPress={onPress}
      isUserSelected={isUserSelected}
      testID="user-card-test-id"
    />
  );

describe('components >> user card ', () => {
  beforeEach(() => onPress.mockClear());

  describe('UserCard without profile image', () => {
    test('Should render user card correctly with no profile image', () => {
      const { toJSON } = setup();
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find username', () => {
      const { getByText } = setup();
      expect(getByText(username)).toBeDefined();
    });

    test('Should fire on press action', () => {
      const { getByText } = setup();
      fireEvent.press(getByText(username));
      expect(onPress).toBeCalledTimes(1);
    });
  });

  describe('UserCard with profile image', () => {
    beforeEach(() => (avatarUrl = 'https://picsum.photos/id/237/200/300'));
    test('Should render user card correctly with profile image', () => {
      const { toJSON } = setup();
      expect(toJSON()).toMatchSnapshot();
    });

    test('Should find username', () => {
      const { getByText } = setup();
      expect(getByText(username)).toBeDefined();
    });

    test('Should fire on press action', () => {
      const { getByText } = setup();
      fireEvent.press(getByText(username));
      expect(onPress).toBeCalledTimes(1);
    });
  });
});
