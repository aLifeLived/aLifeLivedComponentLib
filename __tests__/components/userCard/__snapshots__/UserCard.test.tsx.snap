// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> user card  UserCard with profile image Should render user card correctly with profile image 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Array [
            Object {
              "borderRadius": 96,
              "height": 50,
              "width": 50,
            },
          ],
        ]
      }
    >
      <FastImageView
        iconBorderRadius="xXXlarge"
        resizeMode="cover"
        source={
          Object {
            "priority": "high",
            "uri": "https://picsum.photos/id/237/200/300",
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "marginLeft": 8,
            "width": "80%",
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#282828",
              "display": "flex",
              "fontFamily": "SFUIDisplay-Regular",
              "fontSize": 18,
              "lineHeight": 20,
            },
          ]
        }
      >
        testUsername
      </Text>
    </View>
  </View>
</View>
`;

exports[`components >> user card  UserCard without profile image Should render user card correctly with no profile image 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <View
      iconBorderRadius="xXXlarge"
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 96,
            "borderWidth": 1,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "padding": 16,
            "width": 50,
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={16}
        bbWidth={16}
        fill="#000"
        focusable={false}
        height={16}
        meetOrSlice={0}
        minX={0}
        minY={0}
        secondaryFill={null}
        secondaryOpacity={0.4}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {},
            Object {
              "flex": 0,
              "height": 16,
              "width": 16,
            },
          ]
        }
        vbHeight={512}
        vbWidth={448}
        width={16}
      >
        <RNSVGGroup
          fill={4278190080}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            fill={4278190080}
            fillOpacity={1}
            propList={
              Array [
                "fill",
                "fillOpacity",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "marginLeft": 8,
            "width": "80%",
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#282828",
              "display": "flex",
              "fontFamily": "SFUIDisplay-Regular",
              "fontSize": 18,
              "lineHeight": 20,
            },
          ]
        }
      >
        testUsername
      </Text>
    </View>
  </View>
</View>
`;
